File: tst/layer-hier.c
#include "pico.h"
#include "../check.h"

/* Scene-graph smoke test: one child attached to world.
 *
 * Creates an empty layer under "world", draws a red square into it,
 * and relies on auto-composite (present walk) to blit it onto world.
 * No explicit pico_output_draw_layer call.
 */
int main (void) {
    pico_init(1);

    pico_set_effect_color(PICO_COLOR_BLACK);
    pico_output_clear();

    pico_layer_empty("world", "L", 1, (Pico_Rel_Dim){'!', {8, 8}}, NULL);

    const char* old = pico_set_layer("L");
    pico_set_scene_clear(0);
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.3, 0.3, 0.4, 0.4}, PICO_ANCHOR_C});
    pico_set_effect_color(PICO_COLOR_WHITE);
    pico_output_clear();
    pico_set_pencil_color(PICO_COLOR_RED);
    pico_output_draw_rect(
        (Pico_Rel_Rect){'%', {.5, .5, 0.5, 0.5}, PICO_ANCHOR_C}
    );

    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-01");

    pico_layer_image("world", "img", "../res/open.png");
    old = pico_set_layer("img");
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.7, 0.7, 0.4, 0.4}, PICO_ANCHOR_C});

    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-02");

    static Pico_Color buf[4] = {
        {255, 0, 0, 255}, {0, 255, 0, 255},
        {0, 0, 255, 255}, {255, 255, 0, 255},
    };
    pico_layer_pixmap("world", "buf", (Pico_Abs_Dim){2, 2}, buf);
    old = pico_set_layer("buf");
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.7, 0.3, 0.4, 0.4}, PICO_ANCHOR_C});

    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-03");

    pico_set_pencil_color(PICO_COLOR_GREEN);
    pico_layer_text("world", "txt", 10, "hello");
    old = pico_set_layer("txt");
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.3, 0.7, 0.4, 0.4}, PICO_ANCHOR_C});

    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-04");

    ///////////////////////////////////////////////////////////////////////////

    // transparent overlay with yellow background
    pico_layer_empty("world", "over", 1, (Pico_Rel_Dim){'!', {500, 500}}, NULL);
    old = pico_set_layer("over");
    pico_set_effect_color((Pico_Color){0xFF, 0xFF, 0x00, 0xFF});
    pico_output_clear();
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C});
    pico_set_effect_alpha(0x80);
    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-05");

    // sub-layers: crop blue and green from "buf", parented under "over"
    pico_layer_sub("over", "blue", "buf",
        &(Pico_Rel_Rect){'!', {0, 1, 1, 1}, PICO_ANCHOR_NW});
    old = pico_set_layer("blue");
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.20, 0.9, 0.10, 0.10}, PICO_ANCHOR_C});

    pico_layer_sub("over", "green", "buf",
        &(Pico_Rel_Rect){'!', {1, 0, 1, 1}, PICO_ANCHOR_NW});
    pico_set_layer("green");
    pico_set_scene_dst((Pico_Rel_Rect){'%', {0.40, 0.9, 0.10, 0.10}, PICO_ANCHOR_C});

    pico_set_layer(old);
    pico_output_present(1);
    _pico_check("layer-hier-06");

    pico_init(0);
    return 0;
}

File: lua/tst/layer-hier.lua
require 'pico.check'

-- Scene-graph smoke test: children attached to root.
-- Relies on auto-composite (present walk) to blit onto root.
-- No explicit pico.output.draw.layer call.

pico.init(true)

pico.set.effect { color='black' }
pico.output.clear()

pico.layer.empty { up="world", key="L", dim={'!', w=8, h=8} }

local old = pico.set.layer("L")
pico.set.scene({
    target = {'%', x=0.3, y=0.3, w=0.4, h=0.4, anchor='C'},
    clear = false,
})
pico.set.effect { color='white' }
pico.output.clear()
pico.set.pencil { color='red' }
pico.output.draw.rect(
    {'%', x=0.5, y=0.5, w=0.5, h=0.5, anchor='C'}
)

pico.set.layer(old)
pico.output.present()
pico.check("layer-hier-01")

pico.layer.image { up="world", key="img", path="../../res/open.png",
    target={'%', x=0.7, y=0.7, w=0.4, h=0.4, anchor='C'} }
pico.output.present()
pico.check("layer-hier-02")

pico.layer.pixmap { up="world", key="buf", pixels={
    {{r=255, g=0,   b=0,   a=255}, {r=0, g=255, b=0,   a=255}},
    {{r=0,   g=0,   b=255, a=255}, {r=255, g=255, b=0, a=255}},
}, target={'%', x=0.7, y=0.3, w=0.4, h=0.4, anchor='C'} }
pico.output.present()
pico.check("layer-hier-03")

pico.set.pencil { color='green' }
pico.layer.text { up="world", key="txt", dim={'!', h=10}, text="hello",
    target={'%', x=0.3, y=0.7, w=0.4, h=0.4, anchor='C'} }
pico.output.present()
pico.check("layer-hier-04")

---------------------------------------------------------------------------

-- transparent overlay with yellow background
pico.layer.empty { up="world", key="over", clear=false, dim={'!', w=500, h=500} }
old = pico.set.layer("over")
pico.set.effect { color={r=0xFF, g=0xFF, b=0x00}, alpha=0x80 }
pico.output.clear()
pico.set.scene({
    target = {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'},
})
pico.set.layer(old)
pico.output.present()
pico.check("layer-hier-05")

-- sub-layers: crop blue and green from "buf", parented under "over"
pico.layer.sub { up="over", key="blue", sup="buf",
    crop={'!', x=0, y=1, w=1, h=1, anchor='NW'},
    target={'%', x=0.20, y=0.9, w=0.10, h=0.10, anchor='C'} }

pico.layer.sub { up="over", key="green", sup="buf",
    crop={'!', x=1, y=0, w=1, h=1, anchor='NW'},
    target={'%', x=0.40, y=0.9, w=0.10, h=0.10, anchor='C'} }

pico.output.present()
pico.check("layer-hier-06")

pico.init(false)
